---
title: "Merge_plots"
author: "Robbe Fonteyn"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(Seurat)
```

```{r}
# Plot of the merged object
path <- "C:/Users/Robbe Fonteyn/OneDrive - UGent/AJ 2024-2025/Master's_Dissertation/AML-CITEseq/Data/SAB_Full.rds"
merged_SAB <- readRDS(path)
```

```{r}
merged_SAB <- FindVariableFeatures(merged_SAB, selection.method = "vst", nfeatures = 2000)
```
```{r}
merged_SAB <- ScaleData(merged_SAB, features = rownames(merged_SAB))
```

```{r}
SAB_meta_data <- merged_SAB@meta.data
SAB_meta_data
```

```{r}
colnames(SAB_meta_data)
```


```{r}
library(dplyr)

SAB_meta_data <- SAB_meta_data %>%
  mutate(Patient = coalesce(Patient.x, Patient.y, Patient)) %>%
  select(-Patient.x, -Patient.y)  # Remove the old columns

```


```{r}
# Plot of the merged object
merged_SAB <- RunPCA(merged_SAB, npcs = 30)

# UMAP plot
merged_SAB <- RunUMAP(merged_SAB, dims = 1:30)

# tSNE plot
merged_SAB <- RunTSNE(merged_SAB, dims = 1:30)
```



```{r}
library(dplyr)
library(ggplot2)
DimPlot(merged_SAB, reduction = "umap", label = TRUE) + ggtitle("UMAP Clustering")

# tSNE plot
DimPlot(merged_SAB, reduction = "tsne", label = TRUE) + ggtitle("tSNE Clustering")

# Feature plot for a gene (e.g., "CD34")
FeaturePlot(merged_SAB, features = "CD34", reduction = "umap")

# Violin plot for gene expression across clusters
VlnPlot(merged_SAB, features = "CD34", group.by = "seurat_clusters")

# Ridge plot for gene expression
RidgePlot(merged_SAB, features = "CD34", group.by = "seurat_clusters")

# Heatmap of top differentially expressed genes
top_markers <- FindAllMarkers(merged_SAB)
top10_markers <- top_markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)
DoHeatmap(merged_SAB, features = top10_markers$gene)
```

```{r}
head(merged_SAB@meta.data)
```


```{r}
# Get some annotation on the Dimplot
# change assay to ADT
DefaultAssay(merged_SAB) <- "ADT"
getwd()
DimPlot(merged_SAB, reduction="umap", label=TRUE) + ggtitle("UMAP Clustering with celltype")
#ggsave("UMAP_clustering_with_celltype.png", width=1600, height=900, units="px", scale=2.5)

```

```{r}
FeaturePlot(merged_SAB, features="predicted.celltype.l2.score", reduction="umap") +
  scale_color_viridis_c(option="magma") +  # Use a good color scale
  ggtitle("UMAP Clustering with Gradient of Cell Type Score")

```

```{r}
FeaturePlot(merged_SAB, features="predicted.celltype.l2.score", reduction="umap") +
  scale_color_gradientn(colors=c("blue", "green", "yellow", "red")) +
  ggtitle("UMAP Clustering with Gradient of Cell Type Score") +
  theme_minimal()
```



```{r}
filtered_genes <- top10_markers %>%
  filter(avg_log2FC > 2) %>%
  pull(gene) %>%
  unique()
length(filtered_genes)
```


```{r}
DoHeatmap(merged_SAB, features = filtered_genes)
```


```{r}
metadata <- merged_SAB@meta.data

# Custom violin plot
ggplot(metadata, aes(x = as.factor(seurat_clusters), y = nFeature_RNA, fill = as.factor(seurat_clusters))) +
  geom_violin() +
  theme_minimal() +
  labs(x = "Cluster", y = "Number of Features") +
  ggtitle("Feature Count Distribution")
```

```{r}
library(dplyr)

# Define a threshold for high log2FC (adjust as needed)
log2FC_threshold <- 2  # Change this to 1.5 or another threshold if needed

# Filter for significantly upregulated genes with high log2FC
highly_expressed_genes <- top10_markers %>%
  filter(p_val_adj < 0.05, avg_log2FC > log2FC_threshold) %>%  # Keep only highly upregulated genes
  pull(gene) %>%
  unique()  # Get unique gene names

# Check how many genes are left
length(highly_expressed_genes)

```
```{r}
features_RNA <- rownames(merged_SAB[["RNA"]])
features_ADT <- rownames(merged_SAB[["ADT"]])

DefaultAssay(merged_SAB) <- "ADT"
merged_SAB <- FindVariableFeatures(merged_SAB, selection.method = "vst", nfeatures = 50, assay= "ADT")
head(VariableFeatures(merged_SAB))  # Show the top variable features
```

```{r}
#Plot highly variable features
VariableFeaturePlot(merged_SAB)
top_features <- head(VariableFeatures(merged_SAB), 10)  # Select top 10
LabelPoints(VariableFeaturePlot(merged_SAB), points = top_features, repel = TRUE)
```



